src/webserver/routing.rs

```diff
@@ -0,0 +1,496 @@
+use awc::http::Uri;
+use std::ffi::OsStr;
+use std::path::PathBuf;
+use RoutingAction::{Error, Execute, NotFound, Redirect, Serve};
+
+const INDEX: &'static str = "index.sql";
+const NOT_FOUND: &'static str = "404.sql";
+const EXECUTION_EXTENSION: &'static str = "sql";
+const FORWARD_SLASH: &'static str = "/";
+
+#[derive(Debug, PartialEq)]
+pub enum RoutingAction {
+    Error(String),
+    Execute(PathBuf),
+    NotFound(PathBuf),
+    Redirect(Uri),
+    Serve(PathBuf),
+}
+
+pub trait ExecutionStore {
+    async fn contains(&self, path: &PathBuf) -> bool;
+}
+
+pub trait RoutingConfig {
+    fn prefix(&self) -> &Uri;
+}
+
+pub async fn calculate_route<T, C>(uri: Uri, store: &T, config: &C) -> RoutingAction
+where
+    T: ExecutionStore,
+    C: RoutingConfig,
+{
+    match check_uri(&uri, config) {
+        Ok(path) => match path.clone().extension() {
+            None => calculate_route_without_extension(&uri, path, store).await,
+            Some(extension) => calculate_route_with_extension(path, extension, store).await,
+        },
+        Err(action) => action,
+    }
+}
+
+fn check_uri<C>(uri: &Uri, config: &C) -> Result<PathBuf, RoutingAction>
+where
+    C: RoutingConfig,
+{
+    if uri.path().starts_with(config.prefix().path()) {
+        let mut result = String::from("/");
+        result.push_str(
+            uri.path()
+                .strip_prefix(config.prefix().path())
+                .expect("Unable to remove expected prefix from path"),
+        );
+
+        Ok(PathBuf::from(result))
+    } else {
+        Err(Redirect(config.prefix().clone()))
+    }
```

Comment:
Instead of first checking with `starts_with` and then striping with `strip_prefix`, we should directly match on the result of strip_prefix:

```suggestion
    if let Some(stripped) = uri.path().strip_prefix(config.prefix().path()) {
        Ok(PathBuf::from(format!("/{}", stripped)))
    } else {
        Err(Redirect(config.prefix().clone()))
    }
```